fontFeatures/feeLib/ClassDefinition.py

```diff
@@ -108,56 +108,129 @@
+CONJUNCTOR: "&" | "|" | "-"
+primary: glyphselector | conjunction | predicate | negated_predicate | primary_paren
+primary_paren: "(" primary ")"
+conjunction: primary CONJUNCTOR primary
 
-conjunction = glyphselector:l ws ('&'|'|'|'-'):conjunction ws primary:r -> {'conjunction': {"&":"and","|":"or","-":"subtract"}[conjunction], 'left': l, 'right': r}
-
-DefineClass_Args = classname:c ws '=' ws definition:d -> (c,d)
-definition = primary:g predicate*:p -> (g,p)
+"""
 
-DefineClassBinned_Args = classname:c '[' <letter+>:metric ws "," ws <digit+>:bincount ']' ws '=' ws definition:d -> (metric, bincount, c,d)
+DefineClass_GRAMMAR = """
+?start: action
+action: CLASSNAME "=" primary
+"""
 
-ShowClass_Args = glyphselector:g -> (g,)
+DefineClassBinned_GRAMMAR = """
+?start: action
+action: CLASSNAME "[" METRIC "," NUMBER "]" "=" primary
 """
 
-VERBS = ["DefineClass", "ShowClass", "DefineClassBinned"]
+PARSEOPTS = dict(use_helpers=True)
+VERBS = ["DefineClass", "DefineClassBinned"]
 
-class DefineClass:
-    @classmethod
-    def action(self, parser, classname, definition):
-        glyphs = self.resolve_definition(parser, definition[0])
-        predicates = definition[1]
-        for p in predicates:
-            glyphs = list(filter(lambda x: self.meets_predicate(x, p, parser), glyphs))
+class DefineClass(HelperTransformer):
+    def _add_glyphs_to_named_class(self, glyphs, classname):
         try:
-            parser.fontfeatures.namedClasses[classname["classname"]] = tuple(glyphs)
+            self.parser.fontfeatures.namedClasses[classname] = tuple(glyphs)
         except ValueDuplicationError as e:
             import warnings
             warnings.warn("Could not define class %s as it contains the same"
                     " glyphs as an existing class %s" % (classname["classname"],
-                        parser.fontfeatures.namedClasses.inverse[tuple(glyphs)]))
+                        self.parser.fontfeatures.namedClasses.inverse[tuple(glyphs)]))
+    
+    def has_glyph_predicate(self, args):
+        glyphre, withs = args
+        value = {"replace": re.compile(glyphre.value[1:-1]), "with": withs.value}
+        return {"predicate": "hasglyph", "value": value, "inverted": False}
+
+    def has_anchor_predicate(self, args):
+        (barename,) = args
+        return {"predicate": "hasanchor", "value": barename.value, "inverted": False}
+
+    def category_predicate(self, args):
+        (barename,) = args
+        return {"predicate": "category", "value": barename.value, "inverted": False}
+
+    def _get_metrics(self, glyph, metric=None):
+        metrics = get_glyph_metrics(self.parser.font, glyph)
+        if metric is not None:
+            if metric not in testvalue_metrics:
+                raise ValueError("Unknown metric '%s'" % metric)
+            else:
+                return metrics[metric]
+        else:
+            return metrics
+
+    def metric_comparison(self, args):
+        (metric, comparator, comp_value) = args
```

Comment:
At this point I am starting to think that I would like to "spread" the arguments if possible and if there isn't a good reason why not. Like, the whole point of function arguments is that *they're the arguments to your function*. Passing an array of "arguments to your function" *as an argument to your function* just seems wrong.

Reply:
The only case it could be right is when the array is a predefined structure of related values that always stay together and get passed around multiple places as a set. It doesn't look like that's the case here, I agree this seems wrong.

Reply:
It's a Lark-ism , see my comment below

The main benefit to keeping it is that Lark users will immediately know what to do. It is one of the most popular parsers for Python so...

Reply:
All right. We can probably get used to it. If it really gets to me we can change it later.